lib/bamboo/adapters/smtp_adapter.ex

```diff
@@ -0,0 +1,265 @@
+defmodule Bamboo.SMTPAdapter do
+  @moduledoc """
+  Sends email using SMTP protocol.
+
+  Use this adapter to send emails through SMTP protocol. Requires that some
+  settings are set in the config. See the example section below.
+
+  ## Example config
+
+      # In config/config.exs, or config.prod.exs, etc.
+      config :my_app, MyApp.Mailer,
+        adapter: Bamboo.SMTPAdapter,
+        server: "smtp.domain",
+        port: 1025,
+        username: "your.name@your.domain"
+        password: "pa55word",
+        tls: :if_available, # can be `:always` or `:never`
+        ssl: :false, # can be `:true`
+        retries: 1
+
+      # Define a Mailer. Maybe in lib/my_app/mailer.ex
+      defmodule MyApp.Mailer do
+        use Bamboo.Mailer, otp_app: :my_app
+      end
+  """
+
+  @behaviour Bamboo.Adapter
+
+  require Logger
+
+  @required_configuration [:server, :port, :username, :password]
+  @default_configuration %{tls: :if_available, ssl: :false, retries: 1, transport: :smtp}
+
+  defmodule SMTPError do
+    defexception [:message]
+
+    def exception({reason, detail}) do
+      message = """
+      There was a problem sending the email through SMTP.
+
+      The error is #{inspect reason}
+
+      More detail below:
+
+      #{inspect detail}
+      """
+
+      %SMTPError{message: message}
+    end
+  end
+
+  def deliver(email, config) do
+    gen_smtp_config =
+      config
+      |> to_gen_smtp_server_config
+
+    email
+    |> to_gen_smtp_message
+    |> :gen_smtp_client.send_blocking(gen_smtp_config)
+    |> handle_response
+  end
+
+  @doc false
+  def handle_config(config) do
+    config
+    |> check_required_configuration
+    |> put_default_configuration
+  end
+
+  defp handle_response({:error, reason, detail}) do
```

Comment:
Oh and that's a good point about figuring out what is exported and what's not, but I think most people will usually check the docs and not the source code itself. Also, docs tags are usually added to public functions in libraries which makes it easier to find them. I hope that helps :)